src/utils/dashboardModelUtils.ts

```diff
@@ -0,0 +1,142 @@
+import { RootNodeModel, SectionNodeModel, WidgetModel } from '../interfaces/nodeModels';
+
+
+export function removeWidgetFromState(contents: any, id: string) {
+  // Helper to remove a widget from contents that get passed in
+
+  // This is composed of 2 Steps:
+
+  /*
+      Step 1:
+        - Find the section that the widget is contained in
+        - Remove it from the section's children
+        - Adjust the relative sizing of the section's remaining children
+        - If the sections children are 0 or 1, we need to do some adjustments
+        - If the above is the case call removeSection to handle this
+
+      Step 2:
+        - Remove the Widget from the Widgets list
+  */
+
+  for(let section in contents.sections) {
+    if(contents.sections[section].children.includes(id)) {
+
+      // Call helper to remove the child from the section and redistribute
+      // sizing amongst the remaining children
+      contents = removeChildAndReDistributeSizing(contents, section, id);
+
+      // IF remaining children is 0 or 1 we have to do some extra stuff
+      // This will be handled by the 'removeSection' function
+      if(contents.sections[section].children.length <= 1) {
+        contents = removeSection(contents, contents.widgets[id].parentId);
+      }
+
+      // Given the assumption that a widget can only be in one section
+      // We break to save some time
+      break;
+    }
+
+  }
+
+  // Finally remove the widget
+  delete contents.widgets[id];
```

Comment:
Looking into this, contents here is actually being mutated which isn't what we want

Unfortunately because redux should be immutable, we should *always* return a new copy of contents and never modify the original contents passed in.

In you need to do the following:

```
let rootSection = contents.rootSection  // not sure if this is actually needed but whatever
let sections = {...contents.sections}
let widgets = {...contents.widgets}
```

and then modify those variables instead of the ones contained in `contents.`
and finally return

```
return {
  rootSection: rootSection,
  sections: sections,
  widgets: widgets
} as DashboardContentsModel
```

I found this article helpful: https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns